File: sdk/typescript/src/live-server.test.ts
/**
 * Real-server conformance for the TypeScript SDK.
 *
 * The existing suite is hermetic and tests the client against its own idea of
 * the protocol. That cannot catch the failure that actually breaks an
 * integration: the SDK and the Rust server disagreeing about a field name, a
 * type, or what a verdict means. This one talks to a running Graphite.
 *
 *   GRAPHITE_URL=http://127.0.0.1:7331 GRAPHITE_API_KEY=... npm test
 *
 * Skipped when GRAPHITE_URL is unset, so `npm test` stays offline by default.
 */
import { test, skip } from "node:test";
import assert from "node:assert/strict";
import { GraphiteClient } from "./client.js";
import { computeContentHash } from "./auditbind.js";
import type { VerificationInput } from "./types.js";

const URL = process.env.GRAPHITE_URL;
const KEY = process.env.GRAPHITE_API_KEY;

function client() {
  return new GraphiteClient({ baseUrl: URL!, apiKey: KEY });
}

function transfer(destination: string): VerificationInput {
  return {
    proposed_intent: {
      intent_type: "transfer",
      raw_natural_language: "send SOL",
      confidence_of_parse: 0.95,
    },
    program_id: "11111111111111111111111111111111",
    protocol_version: "1.0.0",
    instruction_discriminator: "02000000",
    account_addresses: [
      "7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU",
      destination,
    ],
    wallet_profile: "Gaming",
    compute_units: 150,
    account_writes: 2,
    cpi_hops: 0,
  } as VerificationInput;
}

const live = URL ? test : skip;

live("a verdict deserializes with every field the gate needs", async () => {
  const r = await client().verify(
    transfer("8qbHbw2BbbTHBW1sbeqakYXVKRQM8Ne7pLK7m6CVfeR"),
  );
  assert.ok(r.content_hash, "content_hash is required for AuditBind");
  assert.ok(r.audit_trail_id, "audit_trail_id missing");
  assert.equal(r.layers?.length, 8, "all eight layers must be reported");
  assert.ok(typeof r.confidence === "number" && r.confidence > 0);
  // The invariant the schema calls load-bearing: these two can never disagree.
  assert.equal(
    r.policy_verdict === "Approved",
    r.approved,
    `policy_verdict ${r.policy_verdict} contradicts approved=${r.approved}`,
  );
});

live("a BLOCK survives the round trip as a block, not an error", async () => {
  // Unspendable destination: the System Program itself.
  const r = await client().verify(transfer("11111111111111111111111111111111"));
  assert.equal(r.approved, false, "the server blocks this; the SDK reported approved");
  assert.equal(r.risk_verdict.status, "Blocked");
  assert.ok(
    r.risk_verdict.findings.some((f) => f.pattern === "UnspendableDestination"),
    `risk findings did not deserialize: ${JSON.stringify(r.risk_verdict.findings)}`,
  );
});

live("AuditBind reproduces the server's content_hash", async () => {
  const input = transfer("8qbHbw2BbbTHBW1sbeqakYXVKRQM8Ne7pLK7m6CVfeR");
  const r = await client().verify(input);
  const local = computeContentHash({
    programId: input.program_id,
    instructionDiscriminator: input.instruction_discriminator,
    accountAddresses: input.account_addresses,
  });
  assert.equal(
    local,
    r.content_hash,
    "cross-language hash mismatch — the TOCTOU check would abort on every legitimate transaction",
  );
});

live("an unauthenticated client is rejected by a keyed server", async () => {
  if (!KEY) return;
  const anon = new GraphiteClient({ baseUrl: URL! });
  await assert.rejects(
    () => anon.verify(transfer("8qbHbw2BbbTHBW1sbeqakYXVKRQM8Ne7pLK7m6CVfeR")),
    "an unauthenticated verify succeeded against a keyed server",
  );
});

File: sdk/go/live_server_test.go
//go:build liveserver

// Real-server conformance for the Go SDK.
//
// The existing tests use httptest handlers, which prove the client speaks its
// own idea of the protocol. They cannot catch the case that actually breaks an
// integration: the SDK and the Rust server disagreeing about a field name, a
// type, or what a verdict means. This suite talks to a running Graphite.
//
//	go test -tags liveserver ./...
//
// with GRAPHITE_URL and GRAPHITE_API_KEY set. Skipped otherwise, so the default
// suite stays hermetic.
package graphite

import (
	"os"
	"testing"
)

func liveClient(t *testing.T) *Client {
	t.Helper()
	url := os.Getenv("GRAPHITE_URL")
	if url == "" {
		t.Skip("GRAPHITE_URL not set — skipping real-server conformance")
	}
	return NewClientWithAPIKey(url, os.Getenv("GRAPHITE_API_KEY"))
}

func transferInput(dest string) *VerificationInput {
	return &VerificationInput{
		ProposedIntent: ProposedIntent{
			IntentType:         "transfer",
			RawNaturalLanguage: "send SOL",
			ConfidenceOfParse:  0.95,
		},
		ProgramID:                "11111111111111111111111111111111",
		ProtocolVersion:          "1.0.0",
		InstructionDiscriminator: "02000000",
		AccountAddresses:         []string{"7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU", dest},
		WalletProfile:            "Gaming",
		ComputeUnits:             150,
		AccountWrites:            2,
		CPIHops:                  0,
	}
}

// The server is reachable and reports itself healthy.
func TestLiveHealth(t *testing.T) {
	c := liveClient(t)
	if err := c.Health(); err != nil {
		t.Fatalf("health failed against a real server: %v", err)
	}
}

// A verdict deserializes completely. A field the SDK silently drops is how an
// integration ends up gating on a zero value.
func TestLiveVerifyDeserializesEveryFieldTheGateNeeds(t *testing.T) {
	c := liveClient(t)
	r, err := c.Verify(transferInput("8qbHbw2BbbTHBW1sbeqakYXVKRQM8Ne7pLK7m6CVfeR"))
	if err != nil {
		t.Fatalf("verify failed: %v", err)
	}
	if r.ContentHash == "" {
		t.Error("content_hash is empty — AuditBind cannot bind the transaction without it")
	}
	if r.AuditTrailID == "" {
		t.Error("audit_trail_id is empty")
	}
	if len(r.Layers) != 8 {
		t.Errorf("expected 8 layer reports, got %d", len(r.Layers))
	}
	if r.Confidence <= 0 {
		t.Errorf("confidence did not deserialize: %v", r.Confidence)
	}
	if r.PolicyVerdict == "" {
		t.Error("policy_verdict is empty")
	}
	// The invariant the schema calls load-bearing.
	if (r.PolicyVerdict == "Approved") != r.Approved {
		t.Errorf("policy_verdict %q contradicts approved=%v", r.PolicyVerdict, r.Approved)
	}
}

// The SDK must carry a BLOCK through as a block. A client that reads a
// rejection as an error and retries, or as a zero value and proceeds, is worse
// than no client.
func TestLiveBlockedVerdictSurvivesTheRoundTrip(t *testing.T) {
	c := liveClient(t)
	// A transfer to the System Program: unspendable destination.
	r, err := c.Verify(transferInput("11111111111111111111111111111111"))
	if err != nil {
		t.Fatalf("a blocked verdict must come back as a result, not a transport error: %v", err)
	}
	if r.Approved {
		t.Fatal("the server blocks this transaction; the SDK reported approved")
	}
	if r.RiskVerdict.Status != "Blocked" {
		t.Errorf("risk status did not deserialize: %q", r.RiskVerdict.Status)
	}
	found := false
	for _, f := range r.RiskVerdict.Findings {
		if f.Pattern == "UnspendableDestination" {
			found = true
		}
	}
	if !found {
		t.Errorf("risk findings did not deserialize: %+v", r.RiskVerdict.Findings)
	}
}

// The Go AuditBind must reproduce the Rust core's content_hash for the same
// transaction. If the two languages disagree, the TOCTOU check either never
// passes or never fails — both useless.
func TestLiveAuditBindAgreesWithTheServer(t *testing.T) {
	c := liveClient(t)
	in := transferInput("8qbHbw2BbbTHBW1sbeqakYXVKRQM8Ne7pLK7m6CVfeR")
	r, err := c.Verify(in)
	if err != nil {
		t.Fatalf("verify failed: %v", err)
	}
	local := ComputeContentHash(AuditBindParams{
		ProgramID:                in.ProgramID,
		InstructionDiscriminator: in.InstructionDiscriminator,
		AccountAddresses:         in.AccountAddresses,
	})
	if local != r.ContentHash {
		t.Fatalf("cross-language hash mismatch: go=%s rust=%s", local, r.ContentHash)
	}
}

// Auth is enforced on the real path, not just in a mock.
func TestLiveMissingAPIKeyIsRejected(t *testing.T) {
	url := os.Getenv("GRAPHITE_URL")
	if url == "" {
		t.Skip("GRAPHITE_URL not set")
	}
	if os.Getenv("GRAPHITE_API_KEY") == "" {
		t.Skip("server has no API key configured; nothing to enforce")
	}
	anon := NewClient(url)
	if _, err := anon.Verify(transferInput("8qbHbw2BbbTHBW1sbeqakYXVKRQM8Ne7pLK7m6CVfeR")); err == nil {
		t.Fatal("an unauthenticated verify succeeded against a keyed server")
	}
}
